shared/manifest.go

```diff
@@ -68,27 +71,95 @@ func (m *Manifest) unmarshalAll() error {
 	return nil
 }
 
-// Contains checks whether m contains the path.
-func (m *Manifest) Contains(path string) (bool, error) {
-	if m.imap == nil {
-		if err := m.unmarshalAll(); err != nil {
-			return false, err
-		}
+func trieContains(t interface{}, parts []string) interface{} {
+	if len(parts) == 0 {
+		return t
 	}
-	parts := strings.Split(path, "/")
-	// Split always returns at least one element.
-	// Remove the leading empty part.
-	if parts[0] == "" {
-		parts = parts[1:]
+
+	// t could be nil (e.g. if the previous part does not exist in the map), in which case casting will fail.
+	trie, ok := t.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+	return trieContains(trie[parts[0]], parts[1:])
+}
+
+// ContainsFile checks whether m contains a file path (including directories).
+func (m *Manifest) ContainsFile(path string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
 	}
+
+	path = strings.Trim(path, "/")
+	if path == "" {
+		// Root directory always exists.
+		return true, nil
+	}
+	parts := strings.Split(path, "/")
 	for _, items := range m.imap {
-		if trieContains(items, parts) {
+		if trieContains(items, parts) != nil {
 			return true, nil
 		}
 	}
 	return false, nil
 }
 
+// ContainsTest checks whether m contains a full test URL.
+func (m *Manifest) ContainsTest(testURL string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
+	}
+
+	// URLs in the manifest do not include the leading slash (url_base).
+	testURL = strings.TrimLeft(testURL, "/")
+	path := testURL
+	query := ""
+	if qPos := strings.Index(testURL, "?"); qPos > -1 {
+		path = testURL[:qPos]
+		query = testURL[qPos:]
+	}
+	path = RecoverTestFilename(path)
+	parts := strings.Split(path, "/")
+	// e.g. testURL="/foo/bar/test.any.html?varaint" would become
+	// testURL="foo/bar/test.any.html?varaint"
+	// path="foo/bar/test.any.js"
+	// query="?variant"
+	// parts=["foo", "bar", "test.any.js"]
+	for _, items := range m.imap {
```

Comment:
Not clear to me why the variable is called `items` here. Isn't `imap` a map from e.g. "testharness" to the (unmarshalled) trie for "testharness" tests? (E.g. isn't it `item`?)

Reply:
`imap` is a map from test types (e.g. "testharness") to a (marshalled) trie of test items. Each test file (leaf node) is an "item". I can rename `items` to `trie`, which probably makes more sense.